Fix parsing of ports inside the URI

flokaiser opened the pull request:
without this change, urls like 5.my.s3.cluster were misinterpreted and
  the '5' parsed as the port of the url.

Bug:
* the getline() function puts the entire input string into the destination string, if the delimiter was not found. This leads to the entire host string being fed to the stoi() function to parse the port.

Fix:
* If the portstr is equal to the host string, no ':port' was encountered. => Skip the parsing of the port

Commit:
attempt parsing of port only if colon was found

without this change, urls like 5.my.s3.cluster were misinterpreted and
  the '5' parsed as the port of the url.

Bug:
* the getline() function puts the entire input string into the
  destination string, if the delimiter was not found.
  This leads to the entire host string being fed to the stoi()
  function to parse the port.

Fix:
* applied suggestion by @balamurugana

kobalicek reviewed (approved):
LGTM - needs one more review and possibly a future test case so this doesn't happen again.

harshavardhana reviewed (changes requested):
Please also add a test

balamurugana commented on src/http.cc:
```diff
@@ -146,7 +146,7 @@ Url Url::Parse(std::string value) {
       }
 
-      if (!portstr.empty()) {
+      if (!portstr.empty() && portstr != host) {
```
This is not the right fix. The fix should be like
```diff
diff --git a/src/http.cc b/src/http.cc
index 8af3dfa..a96177b 100644
--- a/src/http.cc
+++ b/src/http.cc
@@ -140,18 +140,22 @@ Url Url::Parse(std::string value) {
   unsigned int port = 0;
   struct sockaddr_in6 dst;
   if (inet_pton(AF_INET6, host.c_str(), &(dst.sin6_addr)) <= 0) {
-    if (host.front() != '[' || host.back() != ']') {
+    if ((host.front() != '[' || host.back() != ']') && utils::Contains(host, ':')) {
       std::stringstream ss(host);
       std::string portstr;
       while (std::getline(ss, portstr, ':')) {
       }
 
       if (!portstr.empty()) {
-        try {
-          port = static_cast<unsigned>(std::stoi(portstr));
+        char* str_end{};
+        const long l = std::strtol(portstr.c_str(), &str_end, 10);
+        size_t length = std::distance(portstr.c_str(), const_cast<const char*>(str_end));
+        if (length == portstr.size()) {
+          if (l < 1 || l > 65535) {
+            return Url{};
+          }
+          port = static_cast<int>(l);
           host = host.substr(0, host.rfind(":" + portstr));
-        } catch (const std::invalid_argument&) {
-          port = 0;
         }
       }
     }
```

flokaiser replied:
If we want to go further, this is not entirely correct either, because afaik this does not allow ipv6 addresses with port in the form [:::]:1337

I wanted to keep the diff small, but in my humble opinion, the entire parser needs some love :)

balamurugana replied:
Couldn't able to understand why it is marked as resolved.

Commit:
https flag is not mirrored from Url() to BaseUrl

this flag assignement was removed in:
  https://github.com/minio/minio-cpp/pull/111

I don't know if it was removed intentional, but without it
the BaseUrl.https flag only reflects the state of how BaseUrl was
constructed with.

balamurugana commented on src/request.cc:
```diff
@@ -162,6 +162,7 @@ BaseUrl::BaseUrl(std::string host, bool https, std::string region)
     return;
   }
 
+  this->https = url.https;
```
This is not correct. `this->https` is set in the constructor definition already.

flokaiser replied:
I have seen that, but the Url class parses the schema to set the https flag in the url class that…
Correct me if I am wrong, but if I am constructing a u=BasicUrl("https://foo.bar", false)
Then u.https is false

balamurugana replied:
Firstly `BaseUrl` accepts `host`. If URL is used, it takes only host and port. Secondly passed argument should override values. The current behavior is correct and this change is not correct.

Karmile replied:
I don't think this is a normal operation, in fact, I happened to experience this problem and have been troubled for a while for using of an HTTP address but the constructed baseURL being the default HTTPS. I cannot understand choosing to pass parameters instead of using URL parsed values. After all, people are sensitive to their own URLs and are not aware of the construction of functions in third-party libraries

balamurugana replied:
The right fix is to fail if URL is passed like `minio-go` and `minio-py`. We should not support URL style as base URL

jiuker commented:
The reported bug is already fixed on main: the false port parse for hosts without a colon was addressed by #245 (host.find(':') guard), and the BaseUrl https propagation from the URL scheme is also in main. The remaining stricter port-range validation is not in main, but the original issue is resolved. Closing; thank you.
